### ISSUE TITLE
Introduce the concept of attributes

### ISSUE THREAD

**Issue by lorepieri8** [lorepieri8/ARC_challenge#5]
We need to achieve a situation in which the code is able to solve a task like 354 by itself, using basic building blocks functions. 
My visualisation of how this could happen is the following (if you notice, that's how my notebook was structured):
A- The AGI look a t_in and t_out train pairs. It computes all the possible statistics on those. For instance: which colors are inside, how many objects, how many regions, etc.

B- The AGI try to find relations between t_out and t_in, by fitting all the statistics and try to find stats which are satisfied in all the pairs. The more "unlikely" the stats are, the better. Then it picks the most unlikely stat which fits all the pairs. In the 354 case, it will find that max("number of pixels in region") is predictive of the color of t_out.

To summarise my thought: Right now we are looping over transformation and composing them, in the hope to strike the f with gives t_out from t_in. For tasks like 354 this is hopeless, cause we miss reasoning over task attributes. So the loop should be extended also over functions OF THE TASK ATTRIBUTES.

In maths, so far we searched for f, such that f(t_in)=t_out.  In general f is the composition of more fs: f(t_in) = f_2(f_1(t_in)), with f_1:t-->t, f_2:t-->t.

But we should look for is f(t_in,A)=t_out, with A attributes of the train in-out pairs.

And a generic f could be f = f_3(f_2(f_1))), in which f_3 may be f_3:A-->A. So some functions don't act on t (on np.arrays), but on attributes.

In the 354 case, this f_3 is max.
